File: week11/2/Prob5.py
from typing import List

class Solution:
    def maximalSquare(self, matrix: List[List[str]]) -> int:
        n_col = len(matrix[0])
        n_row = len(matrix)
        dp = [[0 for _ in range(n_col+1)] for _ in range(n_row+1)]
        result = 0
        for i in range(1, n_row+1):
            for j in range(1, n_col+1):
                if matrix[i-1][j-1] == "1":
                    dp[i][j] = min(dp[i][j-1], dp[i-1][j-1], dp[i-1][j]) + 1
                    result = max(result, dp[i][j])
        
        return result**2

File: week11/2/Prob5.cpp
#include<iostream>
#include<vector>

using namespace std;

class Solution {
public:
    int maximalSquare(vector<vector<char>>& matrix) {
        int rows = matrix.size(), cols = matrix[0].size(), M = 0;
        vector<vector<int>> dp(rows+1, vector<int>(cols+1, 0));
        
        for (int i = 1; i <= rows; i++){
            for (int j = 1; j <= cols; j++){
                if (matrix[i-1][j-1] == '1'){
                    dp[i][j] = min(min(dp[i][j-1], dp[i-1][j]), dp[i-1][j-1]) + 1;
                    M = max(M, dp[i][j]);
                }
            }
        }
        return M*M;   
    }
};
